<a href="https://colab.research.google.com/github/sergeymasl/pandas_cource/blob/main/Indexing%2C_Selecting_%26_Assigning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Indexing, Selecting & Assigning**

Работа по выбору(фильтрации) данных в DataFrame или Series и получению, необходимых в моменте, данных, будет происходит в большинстве операций, которые вы будете совершать в своей работе.
Поэтому одна из первых вещей, которую вам необходимо изучить - это как быстро и эффективно выбрать из всего массива данных ту часть, которая релевантна для вас в моменте.


https://drive.google.com/uc?export=download&id=1z-1idT4mGbOvHgmEPzneqV54EJ1-w7tk

In [2]:
# импорт библиотеки
import pandas as pd
# загрузка датасета для дальнейшей работы
reviews = pd.read_csv("https://drive.google.com/uc?export=download&id=1z-1idT4mGbOvHgmEPzneqV54EJ1-w7tk", index_col=0)
# настройки по отображению строк
pd.set_option('max_rows', 5)

## **Встроенные возможности**

Язык Python уже использует удобные способы индексации данных (например индексы и срезы в строках и списках, ключи в словарях). Pandas вобрал в себя все эти способы, что позволяет легко начать работу с ним.

Рассмотрим DataFrame загруженный выше:

In [5]:
reviews

country                                        description  \
0          Italy  Aromas include tropical fruit, broom, brimston...   
1       Portugal  This is ripe and fruity, a wine that is smooth...   
...          ...                                                ...   
129969    France  A dry style of Pinot Gris, this is crisp with ...   
129970    France  Big, rich and off-dry, this is powered by inte...   

                          designation  points  price           province  \
0                        Vulkà Bianco      87    NaN  Sicily & Sardinia   
1                            Avidagos      87   15.0              Douro   
...                               ...     ...    ...                ...   
129969                            NaN      90   32.0             Alsace   
129970  Lieu-dit Harth Cuvée Caroline      90   21.0             Alsace   

       region_1 region_2    taster_name taster_twitter_handle  \
0          Etna      NaN  Kerin O’Keefe          @kerinokeefe   
1           NaN      NaN     Roger Voss            @vossroger   
...         ...      ...            ...                   ...   
129969   Alsace      NaN     Roger Voss            @vossroger   
129970   Alsace      NaN     Roger Voss            @vossroger   

                                                    title         variety  \
0                       Nicosia 2013 Vulkà Bianco  (Etna)     White Blend   
1           Quinta dos Avidagos 2011 Avidagos Red (Douro)  Portuguese Red   
...                                                   ...             ...   
129969      Domaine Marcel Deiss 2012 Pinot Gris (Alsace)      Pinot Gris   
129970  Domaine Schoffit 2012 Lieu-dit Harth Cuvée Car...  Gewürztraminer   

                      winery  
0                    Nicosia  
1        Quinta dos Avidagos  
...                      ...  
129969  Domaine Marcel Deiss  
129970      Domaine Schoffit  

[129971 rows x 13 columns]

|    | country   | description                                                                                                                                                                                                                         | designation   |   points |   price | province          | region_1   |   region_2 | taster_name   | taster_twitter_handle   | title                                         | variety        | winery              |
|---:|:----------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------|---------:|--------:|:------------------|:-----------|-----------:|:--------------|:------------------------|:----------------------------------------------|:---------------|:--------------------|
|  0 | Italy     | Aromas include tropical fruit, broom, brimstone and dried herb. The palate isn't overly expressive, offering unripened apple, citrus and dried sage alongside brisk acidity.                                                        | Vulkà Bianco  |       87 |     nan | Sicily & Sardinia | Etna       |        nan | Kerin O’Keefe | @kerinokeefe            | Nicosia 2013 Vulkà Bianco  (Etna)             | White Blend    | Nicosia             |
|  1 | Portugal  | This is ripe and fruity, a wine that is smooth while still structured. Firm tannins are filled out with juicy red berry fruits and freshened with acidity. It's  already drinkable, although it will certainly be better from 2016. | Avidagos      |       87 |      15 | Douro             | nan        |        nan | Roger Voss    | @vossroger              | Quinta dos Avidagos 2011 Avidagos Red (Douro) | Portuguese Red | Quinta dos Avidagos |
| ...|...| ...| ...  |       ... |     ...| ... |... |... | ... | ...| ...| ...|...|
| 129969 | France    | A dry style of Pinot Gris, this is crisp with some acidity. It also has weight and a solid, powerful core of spice and baked apple flavors. With its structure still developing, the wine needs to age. Drink from 2015. | nan                           |       90 |      32 | Alsace     | Alsace     |        nan | Roger Voss    | @vossroger              | Domaine Marcel Deiss 2012 Pinot Gris (Alsace)                               | Pinot Gris     | Domaine Marcel Deiss |
| 129970 | France    | Big, rich and off-dry, this is powered by intense spiciness and rounded texture. Lychees dominate the fruit profile, giving an opulent feel to the aftertaste. Drink now.                                                | Lieu-dit Harth Cuvée Caroline |       90 |      21 | Alsace     | Alsace     |        nan | Roger Voss    | @vossroger              | Domaine Schoffit 2012 Lieu-dit Harth Cuvée Caroline Gewurztraminer (Alsace) | Gewürztraminer | Domaine Schoffit     |

In [7]:
reviews.tail(2).to_markdown()

'|        | country   | description                                                                                                                                                                                                              | designation                   |   points |   price | province   | region_1   |   region_2 | taster_name   | taster_twitter_handle   | title                                                                       | variety        | winery               |\n|-------:|:----------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------------------------|---------:|--------:|:-----------|:-----------|-----------:|:--------------|:------------------------|:----------------------------------------------------------------------------|:---------------|:---------------------|\n| 129969 